jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/CallbackCompletionHandler.java

```diff
@@ -0,0 +1,874 @@
+                LOG.debug("handler callback timeout cancelled={} for {}", cancelled, request);
+
+            completion = failure == null ? "succeeded" : "failed with " + failure;
+            ThreadInfo threadInfo = completionThreadInfo = new ThreadInfo(Thread.currentThread());
+            boolean notify = handled == Boolean.FALSE;
+            if (notify)
+                notifyInvalidHandlerReturnValue(request, threadInfo);
+
+            return true;
+        }
+
+        private ThreadInfo getCompletionThreadInfo()
+        {
+            return completionThreadInfo;
+        }
+
+        @Override
+        public void run()
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("handler callback not completed within {} for {}", getHandlerCallbackTimeout(), request);
+
+            Boolean handled = this.handled;
+
+            ThreadInfo handlerThreadInfo = null;
+            // The Handler chain has not returned yet, likely the thread is blocked.
+            if (handled == null)
+            {
+                // The thread info of the thread that blocks.
+                // Only create it if the handler thread is blocked,
+                // otherwise we will have a stack trace possibly
+                // belonging to a different request.
+                handlerThreadInfo = new ThreadInfo(handleThread);
+            }
+            notifyHandlerCallbackNotCompleted(request, handlerThreadInfo);
+
+            if (isCompleteHandlerCallbackAtTimeout())
+                callback.failed(new TimeoutException());
+        }
+
+        @Override
+        public void dump(Appendable out, String indent) throws IOException
+        {
+            Boolean handled = this.handled;
+            ThreadInfo handleThreadInfo = null;
+            if (handled == null)
+                handleThreadInfo = new ThreadInfo(handleThread);
+            String completion = this.completion;
+            ThreadInfo completionThreadInfo = this.completionThreadInfo;
+
+            out.append("handle() result: %s\n".formatted(handled == null ? "pending" : handled));
+            if (handleThreadInfo != null)
+                out.append(indent).append(handleThreadInfo.toString(indent));
+
+            out.append(indent).append("handler callback: %s\n".formatted(Objects.toString(completion, "not completed")));
+            if (completionThreadInfo != null)
+                out.append(indent).append(completionThreadInfo.toString(indent));
+        }
+    }
+
+    private class RequestWrapper extends Request.Wrapper
+    {
+        private final StateInfo stateInfo;
+
+        private RequestWrapper(StateInfo stateInfo)
+        {
+            super(stateInfo.request);
+            this.stateInfo = stateInfo;
+        }
+
+        @Override
+        public void demand(Runnable reader)
+        {
+            DemandCallback demandCallback = new DemandCallback(reader);
+            stateInfo.demandCallback = demandCallback;
```

Comment:
In the case of a demand call from within a demandCallback, this will replace the existing demandCallback in the stateInfo.  If the callback blocks after making the demand, then we won't see that on a dump.

Maybe stateInfo.demandCallback should be a queue of demandCallbacks, where we put here and remove when the callback has fully completed?

Reply:
Reimplemented demands and writes with a queue.